src/Transform/OptimizeMemoryPools.cpp

```diff
@@ -0,0 +1,905 @@
+  // check if the outermost loops that contain them are the same. If they are
+  // the same then they share the same loop nest, return true.
+  if (!lastOpInTopLevelBlock && !firstLROpInTopLevelBlock &&
+      checkOuterLoopsMatch(lastOp, liveRangeFirstOp))
+    return true;
+
+  // Now check the other pair of extremities. If they are in the same loop nest
+  // return true.
+  if (!firstOpInTopLevelBlock && !lastLROpInTopLevelBlock &&
+      checkOuterLoopsMatch(firstOp, liveRangeLastOp))
+    return true;
+
+  // If none of the cases above were met then:
+  // 1. at least one of the extremities in each pair is at top-block level.
+  // or
+  // 2. extremities are in sub-blocks but they do not share a loop nest.
+  // In either case the intersection check must return false.
+  return false;
+}
+
+/// Check that the live range of the secondGetRef does not intersect with
+/// any of the live ranges of the GetRefs in firstGetRefList.
+bool checkLiveRangesIntersect(
+    SmallVector<KrnlGetRefOp, 4> firstGetRefList, KrnlGetRefOp secondGetRef) {
+  // Check that the live range of each individual element in secondGetRefList
+  // is independent from the individual live ranges of the elements
+  // of the firstGetRefList.
+  for (auto firstGetRef : firstGetRefList) {
+    // Fetch the full live range for the first set of getref operations.
+    std::vector<Operation *> liveRangeOpList = getLiveRange(firstGetRef);
+
+    // Get first and last ops for the live range of the secondGetRef.
+    Operation *firstOp = getLiveRangeFirstOp(secondGetRef);
+    Operation *lastOp = getLiveRangeLastOp(secondGetRef);
+
+    // Check if either the first or last ops in the second live range are part
+    // of the first live range.
+    bool firstOpInLiveRange = operationInLiveRange(firstOp, liveRangeOpList);
+    bool lastOpInLiveRange = operationInLiveRange(lastOp, liveRangeOpList);
+    if (firstOpInLiveRange || lastOpInLiveRange)
+      return true;
+
+    // Since firstOp and lastOp are not part of the live range, check whether
+    // the live range is fully contained between firstOp and lastOp. If it is
+    // return true.
+    if (liveRangeIsContained(firstOp, lastOp, liveRangeOpList))
+      return true;
+
+    // Up to this point, the checks we have done allow for ranges to be
+    // considered disjoint even when their extremities are part of the same
+    // loop nest. This means we have to perform an additional check: if the
+    // extremities of the two live ranges share the same loop nest determiend
+    // by `krnl.iterate` ops. If they do then the live ranges intersect.
+    if (liveRangesInSameLoopNest(firstOp, lastOp, liveRangeOpList))
+      return true;
+  }
+
+  // If all getRef live ranges are independent then no intersection exists.
+  return false;
+}
+
+//===----------------------------------------------------------------------===//
+// Rewrite patterns.
+//===----------------------------------------------------------------------===//
+
```

Comment:
Your comments have been very useful. Do you think you could add one here too for each patterns? Say at high level what they do?

I say this is because I see two transforms... and in the future, maybe more of them will come too.

Reply:
Done